### PR TITLE
Graph protein

### PR DESCRIPTION
**Pull request by tanthanh0510**
update model with DeepDTA and DgraphDTA

### CODE REVIEW

File: lib/config.py [hunk 1]
```diff
@@ -14,21 +14,28 @@
     "gat": "DrugGATNet", 
     "ginconv": "DrugGINConvNet",
     "gat_gcn": "DrugGAT_GCN",
+    "embedding": "EmbeddingSmile"
 }
```
Comment (CuongVoThanh) [inline lib/config.py:19]: MAP_GRAPH_EMBEDDING_MODEL -> MAP_DRUG_MODEL

File: lib/config.py [hunk 2]
```diff
@@ -14,21 +14,28 @@
...
     "ginconv": "NodeGINConvNet",
-    "gat_gcn": "NodeGAT_GCN",
+    "gat_gcn": "NodeGAT_GCN",    
+}
+
+MAP_Protein_EMBEDDING_MODEL = {
+    "embedding": "EmbeddingProtein", 
+    "gcn" : "ProteinGCNNet",
+    'embeddingdeep':'EmbeddingProteinDeep'
 }
```
Comment (CuongVoThanh) [inline lib/config.py:32]: Upper case Protein -> PROTEIN, 
Rename MAP_PROTEIN_MODEL

File: lib/config.py [hunk 3]
```diff
@@ -66,25 +73,37 @@ def get_dataset(self, **kwargs):
 
     def get_graph_embedding(self, **kwargs):
         name = self.config["model"]["graph_embedding"]
-        return getattr(graph_embedding, name)(**kwargs)
+        parameter_model = {"n_filters": 32}
+        if self.config["datasets"]["dataset"] == 'davis' and self.config["model"]["graph_embedding"] == 'EmbeddingSmile':
+            parameter_model['in_channels'] = 85
+        else:
+            parameter_model['in_channels'] = 100
+        return getattr(graph_embedding, name)(**parameter_model,**kwargs)
```
Comment (CuongVoThanh) [inline lib/config.py:82]: Move n_filters, in_channels to self.config

File: lib/config.py [hunk 4]
```diff
@@ -66,25 +73,37 @@ def get_dataset(self, **kwargs):
...
     def get_protein_model(self, **kwargs):
         name = self.config["model"]["protein"]
-        return getattr(models_protein, name)(**kwargs)
+        parameter_model = {"n_filters": 32}
+        if self.config["datasets"]["dataset"] == 'davis' and self.config["model"]["protein"] == 'EmbeddingProteinDeep':
+            parameter_model['in_channels'] = 1200
+        else:
+            parameter_model['in_channels'] = 1000
+
+        return getattr(models_protein, name)(**parameter_model,**kwargs)
```
Comment (CuongVoThanh) [inline lib/config.py:96]: Move n_filters, in_channels to self.config

File: lib/config.py [hunk 5]
```diff
@@ -107,7 +126,6 @@ def __contains__(self, item):
...
 
     def __getnum_node(self):
-        dataset = self.get_dataset(mode='train')
-        num_drug = dataset.data.num_drug[0].item()
-        num_protein = dataset.data.num_protein[0].item()
-        return num_drug+num_protein
\ No newline at end of file
+        if self.config["datasets"]["dataset"] == "davis":
+            return 510
+        return 2340
```
Comment (CuongVoThanh) [inline lib/config.py:131]: Refactor this code, 
Can we utilize dataset to calculate 510 or 2340

File: lib/runner.py [hunk 6]
```diff
@@ -70,13 +72,12 @@ def train(self, fold: int = 1) -> tuple:
...
             pbar = tqdm(train_loader, desc=f"Train on epoch {epoch}")
             for i, data in enumerate(pbar):
-                data = data.to(self.device)
                 optimizer.zero_grad()
                 if self.cfg.config["model"]["model"] == "SGDTI":
                     output = model(data, edge_index_drug_protein)
                 else:
                     output = model(data)
-                loss = loss_fn(output, data.y.view(-1, 1).float().to(self.device))
+                loss = loss_fn(output, data[0].y.view(-1, 1).float().to(self.device))
```
Comment (CuongVoThanh) [inline lib/runner.py:80]: Check side effect?

File: lib/runner.py [hunk 7]
```diff
@@ -7,9 +7,11 @@
 
 import torch
 
-from torch_geometric.data import DataLoader
+# from torch_geometric.data import DataLoader
+from torch.utils.data import DataLoader
 
 from utils.evaluation_metrics import mse, ci
+from utils.create_data import collate
```
Comment (CuongVoThanh) [inline lib/runner.py:14]: Is that collate in dataset rather than runner?

File: lib/runner.py [hunk 8]
```diff
@@ -121,15 +122,14 @@ def eval(self,
         with torch.no_grad():
             pbar = tqdm(dataloader, desc=des)
             for data in pbar:
-                data = data.to(self.device)
                 if self.cfg.config["model"]["model"] == "SGDTI":
-                    new_edge = torch.LongTensor(data.edge_index_drug_protein).transpose(1, 0).to(self.device)
+                    new_edge = torch.LongTensor(data[0].edge_index_drug_protein).transpose(1, 0).to(self.device)
```
Comment (CuongVoThanh) [inline lib/runner.py:126]: Check side effect

File: lib/runner.py [hunk 9]
```diff
@@ -121,15 +122,14 @@ def eval(self,
...
                     edge_index_drug_protein = torch.cat((edge_index_drug_protein, new_edge, new_edge[[1,0]]), 1)
                     output = model(data,edge_index_drug_protein)                    
                 else:
                     output = model(data)
                 total_preds = torch.cat((total_preds, output.cpu()), 0)
-                total_labels = torch.cat((total_labels, data.y.view(-1, 1).cpu()), 0)
+                total_labels = torch.cat((total_labels, data[0].y.view(-1, 1).cpu()), 0)
```
Comment (CuongVoThanh) [inline lib/runner.py:132]: Check side effect

File: lib/runner.py [hunk 10]
```diff
@@ -153,9 +153,12 @@ def get_train_dataloader(self, fold: int) -> DataLoader:
         train_dataset = self.cfg.get_dataset(mode='train', fold=fold)
         
         edge_index_drug_protein = train_dataset.edge_index_drug_protein
+
+
```
Comment (CuongVoThanh) [inline lib/runner.py:157]: enter line a lot?

File: main.py [hunk 11]
```diff
@@ -5,15 +5,18 @@
 from lib.experiment import Experiment
 from lib.config import Config
 from lib.runner import Runner
+from lib.load_data.datasets import Datasets
+from torch_geometric.data import DataLoader
+from utils.create_data import collate
```
Comment (CuongVoThanh) [inline main.py:10]: What are lines 8,9,10 use for???????

File: main.py [hunk 12]
```diff
@@ -5,15 +5,18 @@
...
     parser.add_argument("--dataset", choices=["davis", "drugchem", "kiba"], help="choose dataset", required=True)
-    parser.add_argument("--graph_embedding", choices=["gcn", "gat", "ginconv", "gat_gcn"], help="choose graph_embedding for drug", required=True)
+    parser.add_argument("--graph_embedding", choices=["gcn", "gat", "ginconv", "gat_gcn",'embedding'], help="choose graph_embedding for drug", required=True)
     parser.add_argument("--node_embedding", default="fc_layer", choices=["gcn", "gat", "ginconv", "gat_gcn", "fc_layer"], help="choose stacked graph dti", required=True)
-    parser.add_argument("--protein_model", default="EmbeddingProtein", choices=["EmbeddingProtein"], help="choose stacked graph dti")
-    parser.add_argument("--model", default="graphdta" ,choices=["graphdta", "sgdti"], help="choose model", required=True)
+    parser.add_argument("--protein_model", default="embedding", choices=["embedding","gcn",'embeddingdeep'], help="choose stacked graph dti")
+    parser.add_argument("--model", default="graphdta" ,choices=["graphdta", "sgdti",'deep'], help="choose model", required=True)
```
Comment (CuongVoThanh) [inline main.py:19]: rename _deep_ -> _deepdta_

File: main.py [hunk 13]
```diff
@@ -78,4 +81,5 @@ def main():
     return
 
 if __name__ == "__main__":
-    main()
\ No newline at end of file
+    main()
+
```
Comment (CuongVoThanh) [inline main.py:85]: space

File: utils/create_data.py [hunk 14]
```diff
@@ -1,21 +1,96 @@
...
+
+
+def residue_features(residue):
+    res_property1 = [1 if residue in PRO_RES_ALIPHATIC_TABLE else 0, 1 if residue in PRO_RES_AROMATIC_TABLE else 0,
+                     1 if residue in PRO_RES_POLAR_NEUTRAL_TABLE else 0,
+                     1 if residue in PRO_RES_ACIDIC_CHARGED_TABLE else 0,
+                     1 if residue in PRO_RES_BASIC_CHARGED_TABLE else 0]
+    res_property2 = [RES_WEIGHT_TABLE[residue], RES_PKA_TABLE[residue], RES_PKB_TABLE[residue], RES_PKX_TABLE[residue],
+                     RES_PL_TABLE[residue], RES_HYDROPHOBIC_PH2_TABLE[residue], RES_HYDROPHOBIC_PH7_TABLE[residue]]
+    # print(np.array(res_property1 + res_property2).shape)
```
Comment (CuongVoThanh) [inline utils/create_data.py:84]: remove print line

File: utils/create_data.py [hunk 15]
```diff
@@ -41,11 +116,67 @@ def smile_to_graph(smile):
...
\ No newline at end of file
+    return Chem.MolToSmiles(Chem.MolFromSmiles(ligand), isomericSmiles = isomericSmiles)
+
+def PSSM_calculation(aln_file, pro_seq):
+    pfm_mat = np.zeros((len(PRO_RES_TABLE), len(pro_seq)))
+    with open(aln_file, 'r') as f:
+        line_count = len(f.readlines())
+        for line in f.readlines():
+            if len(line) != len(pro_seq):
+                print('error', len(line), len(pro_seq))
```
Comment (CuongVoThanh) [inline utils/create_data.py:134]: use self.info instead print

File: utils/create_data.py [hunk 16]
```diff
@@ -41,11 +116,67 @@ def smile_to_graph(smile):
...
+    target_size = len(target_sequence)
+    contact_file = os.path.join(contact_dir, target_key + '.npy')
+    contact_map = np.load(contact_file)
+    contact_map += np.matrix(np.eye(contact_map.shape[0]))
+    index_row, index_col = np.where(contact_map >= 0.5)
+    for i, j in zip(index_row, index_col):
+        target_edge_index.append([i, j])
+    target_feature = target_to_feature(target_key, target_sequence, aln_dir)
+    target_edge_index = np.array(target_edge_index)
+    return target_size, target_feature, target_edge_index
```
Comment (CuongVoThanh) [inline utils/create_data.py:177]: refactor target_to_graph

File: utils/create_data.py [hunk 17]
```diff
@@ -41,11 +116,67 @@ def smile_to_graph(smile):
...
+
+def collate(data_list):
+    batchA = Batch.from_data_list([data[0] for data in data_list])
+    batchB = Batch.from_data_list([data[1] for data in data_list])
+    return batchA, batchB
```
Comment (CuongVoThanh) [inline utils/create_data.py:182]: rename A,B

File: utils/create_data.py [hunk 18]
```diff
@@ -41,11 +116,67 @@ def smile_to_graph(smile):
...
+                pfm_mat[PRO_RES_TABLE.index(res), count] += 1
+                count += 1
+    pseudocount = 0.8
+    ppm_mat = (pfm_mat + pseudocount / 4) / (float(line_count) + pseudocount)
+    pssm_mat = ppm_mat
+    return pssm_mat
+
+def seq_feature(pro_seq):
+    pro_hot = np.zeros((len(pro_seq), len(PRO_RES_TABLE)))
+    pro_property = np.zeros((len(pro_seq), 12))
```
Comment (CuongVoThanh) [inline utils/create_data.py:150]: what does _12_ represent for?

File: lib/load_data/datasets.py [hunk 19]
```diff
@@ -44,13 +44,14 @@ def __init__(self,
...
+        if self.mode == "test":            
+            self.data_mol, self.data_pro = torch.load(self.file_name_test + ".pt")
+            self.edge_index_drug_protein = torch.load(self.file_name_test + "_edge_index_drug_protein.pt")
         elif self.mode == "train":
-            self.data, self.slices = torch.load(self.files_name_train[self.fold] + ".pt")
+            self.data_mol, self.data_pro = torch.load(self.files_name_train[self.fold] + ".pt")
             self.edge_index_drug_protein = torch.load(self.files_name_train[self.fold] + "_edge_index_drug_protein.pt")
         elif self.mode == "val":
-            self.data, self.slices = torch.load(self.files_name_val[self.fold] + ".pt")
+            self.data_mol, self.data_pro = torch.load(self.files_name_val[self.fold] + ".pt")
```
Comment (CuongVoThanh) [inline lib/load_data/datasets.py:55]: Consider this bug?

File: lib/load_data/datasets.py [hunk 20]
```diff
@@ -142,14 +144,17 @@ def __post_process_data(self, file_name: str, message: str) -> None:
...
-        torch_data = self.convert_to_torch_data(df_testset, message)
-        data, slices = self.collate(list(torch_data))
-        edge_index_drug_protein = torch.LongTensor(data.edge_index_drug_protein).transpose(1, 0)
+        data_list_mol,data_list_pro = self.convert_to_torch_data(df_testset, message,self.dataset)
+        edge_index_drug_protein = torch.LongTensor([i.edge_index_drug_protein for i in data_list_mol]).transpose(1, 0)
         edge_index_drug_protein = torch.cat((edge_index_drug_protein, edge_index_drug_protein[[1,0]]), 1)
     
         torch.save((edge_index_drug_protein), file_name + "_edge_index_drug_protein.pt")
-        torch.save((data, slices), file_name + ".pt")
+        torch.save((data_list_mol, data_list_pro), file_name + ".pt")
```
Comment (CuongVoThanh) [inline lib/load_data/datasets.py:152]: Enter line

File: lib/load_data/datasets.py [hunk 21]
```diff
@@ -173,46 +178,79 @@ def write_to_csv(file_name_test: str,
...
         """ 
-        drugs, proteins, affinity,rows,cols,num_drug,num_protein = list(df['compound_iso_smiles']), list(df['target_sequence']), list(df['affinity']),list(df['rows']),list(df['cols']),list(df['num_drug']),list(df['num_protein'])
-        proteins = [seq_cat(protein) for protein in proteins]
+        msa_path = 'data/' + dataset + '/aln'
+        contac_path = 'data/' + dataset + '/pconsc4'
+        drugs, proteins,proteins_name, affinity,rows,cols,num_drug,num_protein = list(df['compound_iso_smiles']), list(df['target_sequence']), list(df['target_name']), list(df['affinity']),list(df['rows']),list(df['cols']),list(df['num_drug']),list(df['num_protein'])
+        
+        proteins_one_hot = [seq_cat(protein) for protein in proteins]
+
+        proteins_one_hot_deep = [seq_cat(protein,SEQ_DICT,1200 if dataset == "davis" else 1000) for protein in proteins]
```
Comment (CuongVoThanh) [inline lib/load_data/datasets.py:210]: Consider this line code

File: lib/load_data/datasets.py [hunk 22]
```diff
@@ -173,46 +178,79 @@ def write_to_csv(file_name_test: str,
...
+
+        smile_one_hot = [seq_cat(drug,CHARISOSMISET,85 if dataset == "davis" else 100) for drug in drugs]
```
Comment (CuongVoThanh) [inline lib/load_data/datasets.py:212]: Can we do better?

File: lib/load_data/datasets.py [hunk 23]
```diff
@@ -173,46 +178,79 @@ def write_to_csv(file_name_test: str,
...
+
+
+            tar_c_size, tar_features, tar_edge_index = target_to_graph(target_name,target_seq,contac_path,msa_path)            
+
+            target_graph = DATA.Data(x=torch.Tensor(tar_features),
+                                edge_index=torch.LongTensor(tar_edge_index).transpose(1, 0),
+                                y=torch.FloatTensor([labels]))
+
+            target_graph.__setitem__("c_size",tar_c_size)
+            GCNData.__setitem__("c_size",_)
```
Comment (CuongVoThanh) [inline lib/load_data/datasets.py:243]: c_size = _  ?????

File: lib/config.py (hunk 1 above)
Comment (tanthanh0510) [reply lib/config.py:19]: Done

File: lib/config.py (hunk 2 above)
Comment (tanthanh0510) [reply lib/config.py:32]: Done

File: lib/runner.py (hunk 6 above)
Comment (tanthanh0510) [reply lib/runner.py:80]: no effect

File: lib/runner.py (hunk 8 above)
Comment (tanthanh0510) [reply lib/runner.py:126]: no effect

File: lib/runner.py (hunk 9 above)
Comment (tanthanh0510) [reply lib/runner.py:132]: no effect

File: lib/load_data/datasets.py (hunk 23 above)
Comment (tanthanh0510) [reply lib/load_data/datasets.py:243]: done

File: lib/load_data/datasets.py (hunk 22 above)
Comment (tanthanh0510) [reply lib/load_data/datasets.py:212]: how?

File: lib/load_data/datasets.py (hunk 21 above)
Comment (tanthanh0510) [reply lib/load_data/datasets.py:210]: it's the same as the previous version

File: lib/load_data/datasets.py (hunk 20 above)
Comment (tanthanh0510) [reply lib/load_data/datasets.py:152]: done

File: lib/load_data/datasets.py (hunk 19 above)
Comment (tanthanh0510) [reply lib/load_data/datasets.py:55]: Done

File: utils/create_data.py (hunk 17 above)
Comment (tanthanh0510) [reply utils/create_data.py:182]: Done

File: utils/create_data.py (hunk 15 above)
Comment (tanthanh0510) [reply utils/create_data.py:134]: Done

File: utils/create_data.py (hunk 14 above)
Comment (tanthanh0510) [reply utils/create_data.py:84]: Done

File: utils/create_data.py (hunk 18 above)
Comment (tanthanh0510) [reply utils/create_data.py:150]: 12 represent for:
1.  Whether the residue is aliphatic
2. Whether the residue is aromatic
3. Whether the residue is polar neutral
4.  Whether the residue is acidic charged
5.  Whether the residue is basic charged
6. Residue weight
7.  The negative of the logarithm of the dissociation constant for the –COOH group64
8.  The negative of the logarithm of the dissociation constant for the –NH3 group64
9.  The negative of the logarithm of the dissociation constant for any other group in the molecule64
10. The pH at the isoelectric point64
11. Hydrophobicity of residue (pH ¼ 2)65
12.  Hydrophobicity of residue (pH ¼ 7)66
See detail in paper DgraphDTA: https://pubs.rsc.org/en/content/articlepdf/2020/ra/d0ra02297g

File: utils/create_data.py (hunk 16 above)
Comment (tanthanh0510) [reply utils/create_data.py:177]: refactor line 176?

File: main.py (hunk 13 above)
Comment (tanthanh0510) [reply main.py:85]: Done

File: main.py (hunk 12 above)
Comment (tanthanh0510) [reply main.py:19]: done

File: main.py (hunk 11 above)
Comment (tanthanh0510) [reply main.py:10]: for test code!! :))

File: lib/runner.py (hunk 10 above)
Comment (tanthanh0510) [reply lib/runner.py:157]: done

File: lib/runner.py (hunk 7 above)
Comment (tanthanh0510) [reply lib/runner.py:14]: use for create dataloader

File: lib/models/sgdti_model.py [hunk 24]
```diff
@@ -35,17 +36,22 @@ def __init__(self,
 
         self.out = nn.Linear(2, 1)
 
+        self.device = device
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/models/sgdti_model.py:39]: no. class need device to sync input

File: lib/models/sgdti_model.py [hunk 25]
```diff
@@ -35,17 +36,22 @@ def __init__(self,
...
+        
+        embedding_protein = self.proteins(data)
 
         output_1 = self.fc_left(feature_drug, embedding_protein)
         
         with torch.no_grad():
-            drug_id = np.array(data.edge_index_drug_protein)[:,0] #TODO:
-            protein_id = np.array(data.edge_index_drug_protein)[:,1]
+            drug_id = np.array(data[0].edge_index_drug_protein)[:,0] #TODO:
+            protein_id = np.array(data[0].edge_index_drug_protein)[:,1]
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/models/sgdti_model.py:55]: no effect

File: lib/models/models_protein/gcn.py [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+
+        self.relu = nn.ReLU()
+        self.dropout = nn.Dropout(p)
+
+        self.device = device
+        print(device)
+
+    def forward(self, data: tuple) -> torch.Tensor:
+        target = data[1]
+        target = target.to(self.device)
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/models/models_protein/gcn.py:28]: no, i need device in class

File: lib/models/models_protein/gcn.py [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+        self.pro_conv1 = GCNConv(num_features_pro, num_features_pro)
+        self.pro_conv2 = GCNConv(num_features_pro, num_features_pro * 2)
+        self.pro_conv3 = GCNConv(num_features_pro * 2, num_features_pro * 4)
+        self.pro_fc_g1 = torch.nn.Linear(num_features_pro * 4, 1024)
+        self.pro_fc_g2 = torch.nn.Linear(1024, output_dim)
+
+        self.relu = nn.ReLU()
+        self.dropout = nn.Dropout(p)
+
+        self.device = device
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/models/models_protein/gcn.py:23]: i need device

File: lib/models/deep_model.py [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+        self.input_dim = input_dim
+        self.p = p # Dropout Rate
+
+        self.fc1 = nn.Linear(2*self.input_dim, 1024)
+        self.fc2 = nn.Linear(1024, 512)
+        self.out = nn.Linear(512, self.n_output)
+        self.relu = nn.ReLU()
+        self.dropout = nn.Dropout(self.p)
+
+        self.device = device
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/models/deep_model.py:26]: i need device

File: lib/load_data/datasets.py [hunk 29]
```diff
@@ -173,46 +178,79 @@ def write_to_csv(file_name_test: str,
...
                     row += [1]
-                else:
+                else:                    
                     row += [affinity[rows[pair_ind], cols[pair_ind]]]
                     row += [rows[pair_ind], cols[pair_ind], affinity.shape[0], affinity.shape[1]]
                 f.write(','.join(map(str, row)) + '\n')
 
     @staticmethod
-    def convert_to_torch_data(df: pd.DataFrame, des: str) -> list:
+    def convert_to_torch_data(df: pd.DataFrame, des: str,dataset: str) -> list:
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/load_data/datasets.py:200]: Done

File: lib/models/models_protein/embeddingDeep.py [hunk 30]
```diff
@@ -0,0 +1,48 @@
...
+        self.embedding_protein = nn.Embedding(self.num_features_protein+1, self.embed_dim)
+        self.conv_protein_1 = nn.Conv1d(in_channels=self.in_channels, out_channels=self.n_filters, kernel_size=8)
+        self.conv_protein_2 = nn.Conv1d(in_channels=self.n_filters, out_channels=self.n_filters*2, kernel_size=8)
+        self.conv_protein_3 = nn.Conv1d(in_channels=self.n_filters*2, out_channels=self.n_filters*3, kernel_size=8)
+        self.fc1_protein = nn.Linear(FLATTEN_SHAPE, self.output_dim)
+
+        self.relu = nn.ReLU()
+
+        self.device = device
+    def forward(self, data) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/models/models_protein/embeddingDeep.py:32]: Done

File: lib/models/models_protein/embeddingDeep.py [hunk 31]
```diff
@@ -0,0 +1,48 @@
...
+
+        self.embedding_protein = nn.Embedding(self.num_features_protein+1, self.embed_dim)
+        self.conv_protein_1 = nn.Conv1d(in_channels=self.in_channels, out_channels=self.n_filters, kernel_size=8)
+        self.conv_protein_2 = nn.Conv1d(in_channels=self.n_filters, out_channels=self.n_filters*2, kernel_size=8)
+        self.conv_protein_3 = nn.Conv1d(in_channels=self.n_filters*2, out_channels=self.n_filters*3, kernel_size=8)
+        self.fc1_protein = nn.Linear(FLATTEN_SHAPE, self.output_dim)
+
+        self.relu = nn.ReLU()
+
+        self.device = device
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/models/models_protein/embeddingDeep.py:31]: noooo

File: lib/models/models_protein/embedding.py [hunk 32]
```diff
@@ -21,13 +23,13 @@ def __init__(self,
         self.conv_protein_1 = nn.Conv1d(in_channels=1000, out_channels=self.n_filters, kernel_size=8)
         self.fc1_protein = nn.Linear(FLATTEN_SHAPE, self.output_dim)
 
-    def forward(self, target: torch.Tensor) -> torch.Tensor:
+        self.device = device
+    def forward(self, data: tuple) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/models/models_protein/embedding.py:27]: done

File: lib/models/models_protein/embedding.py [hunk 33]
```diff
@@ -9,7 +10,8 @@ def __init__(self,
                 n_filters: int = 32,
                 embed_dim: int = 128,
                 num_features_protein: int = 25,
-                output_dim: int = 128) -> None:
+                output_dim: int = 128,
+                device: None or torch.device = None) -> None:
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/models/models_protein/embedding.py:14]: i neet it

File: lib/models/graphdta_model.py [hunk 34]
```diff
@@ -24,9 +23,13 @@ def __init__(self,
...
-    def forward(self, data: Datasets) -> torch.tensor:
-        drug_features = self.drug(data.batch, data.edge_index, data.x)
-        protein_features = self.proteins(data.target)
+        self.device = device
+
+    def forward(self, data: tuple) -> torch.tensor:
+        batch_drug,edge_index_drug,x_drug = data[0].batch,data[0].edge_index,data[0].x
+        batch_drug,edge_index_drug,x_drug = batch_drug.to(self.device),edge_index_drug.to(self.device),x_drug.to(self.device)
+        drug_features = self.drug(batch_drug,edge_index_drug,x_drug)
+        protein_features = self.proteins(data)
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/models/graphdta_model.py:33]: no,no

File: lib/models/graph_embedding/embedding.py [hunk 35]
```diff
@@ -0,0 +1,50 @@
...
+        self.embedding_smile = nn.Embedding(self.num_features_smile+1, self.embed_dim)
+        self.conv_smile_1 = nn.Conv1d(in_channels=self.in_channels, out_channels=self.n_filters, kernel_size=8)
+        self.conv_smile_2 = nn.Conv1d(in_channels=self.n_filters, out_channels=self.n_filters*2, kernel_size=8)
+        self.conv_smile_3 = nn.Conv1d(in_channels=self.n_filters*2, out_channels=self.n_filters*3, kernel_size=8)
+        self.fc1_smile = nn.Linear(FLATTEN_SHAPE, self.output_dim)
+
+        self.relu = nn.ReLU()
+
+        self.device = device
+    def forward(self, data) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/models/graph_embedding/embedding.py:32]: Done

File: lib/experiment.py [hunk 36]
```diff
@@ -59,7 +59,7 @@ def get_checkpoint_path(self, epoch: int, fold: int):
         return os.path.join(self.models_dirpath, 'fold_{:04d}_model_{:04d}.pt'.format(fold, epoch))
 
     def get_epoch_model(self, epoch: int, fold: int):
-        return torch.load(self.get_checkpoint_path(epoch, fold))['model']
+        return torch.load(self.get_checkpoint_path(epoch, fold),map_location=torch.device("cpu"))['model']
```
[the remark this answers is not in the document]
Comment (tanthanh0510) [reply lib/experiment.py:62]: for load model with cpu

File: lib/load_data/datasets.py (hunk 21 above)
Comment (tanthanh0510) [reply lib/load_data/datasets.py:210]: Done

File: lib/load_data/datasets.py (hunk 22 above)
Comment (tanthanh0510) [reply lib/load_data/datasets.py:212]: Done

**Pull request closed without merging**